src/callbacks/callback_imcomm.cpp

```diff
@@ -132,16 +132,20 @@ void lbann_callback_imcomm::on_epoch_end(model *m) {
     optimizer *opt = w->get_optimizer();
     if (ct_does_quantization(params.ct)) {
       comm->intermodel_sum_matrix(params.error);
+      opt->clear_gradient();
+      auto gradient = opt->get_gradient().Copy();
```

Comment:
I've changed `optimizer::get_gradient()` to return a const reference to a matrix since things can get squirrelly if the data lives on GPU. In particular, someone could think they are changing the gradient by changing the CPU matrix, but it would actually have no effect. Once we have GPU support in Hydrogen, we'll no longer need to maintain separate matrices for CPU and GPU matrices and this concern will be moot.